source/common/modules/markdown-editor/plugins/render-math.ts

```diff
@@ -0,0 +1,167 @@
+
+import { getBlockMathRE, getInlineMathRenderRE } from '../../../../common/regular-expressions'
+import * as CodeMirror from 'codemirror'
+import katex from 'katex'
+
+require('katex/dist/contrib/mhchem.js') // modify katex module
+
+const multilineMathRE = getBlockMathRE()
+const commands = (CodeMirror.commands as any)
+commands.markdownRenderMath = function (cm: CodeMirror.Editor) {
+  // First, find all math elements
+  // We'll only render the viewport
+  const viewport = cm.getViewport()
+  let lines: LineInfo[] = []
+  for (let i = viewport.from; i < viewport.to; i++) {
+    let modeName = cm.getModeAt({ 'line': i, 'ch': 0 }).name
+    let tokenType = cm.getTokenTypeAt({ 'line': i, 'ch': 0 })
+    lines.push(new LineInfo(i, cm.getLine(i), modeName, tokenType))
+  }
+  let equations = findEquations(lines)
+
+  // Now cycle through all new markers and insert them, if they weren't already
+  for (let myMarker of equations) {
+    let cur = cm.getCursor('from')
+    let isMulti = myMarker.curFrom.line !== myMarker.curTo.line
+    if (isMulti && cur.line >= myMarker.curFrom.line && cur.line <= myMarker.curTo.line) {
+      // We're directly in the multiline equation, so don't render.
+      continue
+    } else if (!isMulti && cur.line === myMarker.curFrom.line && cur.ch >= myMarker.curFrom.ch && cur.ch <= myMarker.curTo.ch) {
+      // Again, we're right in the middle of an inline-equation, so don't render.
+      continue
+    }
+
+    // We can only have one marker at any given position at any given time
+    if (cm.getDoc().findMarks(myMarker.curFrom, myMarker.curTo).length > 0) continue
+
+    // Do not render if it's inside a comment (in this case the mode will be
+    // markdown, but comments shouldn't be included in rendering)
+    // Final check to avoid it for as long as possible, as getTokenAt takes
+    // considerable time.
+    let tokenTypeBegin = cm.getTokenTypeAt(myMarker.curFrom)
+    let tokenTypeEnd = cm.getTokenTypeAt(myMarker.curTo)
+    if ((tokenTypeBegin?.includes('comment')) || (tokenTypeEnd?.includes('comment'))) {
+      continue
+    }
+
+    let mathSpan = document.createElement('span')
+    mathSpan.classList.add('preview-math')
+
+    let textMarker = cm.getDoc().markText(
+      myMarker.curFrom, myMarker.curTo,
+      {
+        'className': 'test',
+        'clearOnEnter': true,
+        'replacedWith': mathSpan,
+        'inclusiveLeft': false,
+        'inclusiveRight': false
+      }
+    )
+
+    // Enable on-click closing of rendered Math elements.
+    mathSpan.onclick = (e) => { textMarker.clear() }
+
+    katex.render(myMarker.eq, mathSpan, { throwOnError: false, displayMode: myMarker.displayMode })
+
+    // Now the marker has obviously changed
+    textMarker.changed()
+  }
+}
+
+export class LineInfo {
+  constructor (
+    readonly lineNumber: number,
+    readonly text: string,
+    readonly modeName: string,
+    readonly tokenType?: string
+  ) {}
+}
+
+export class EquationMarker {
```

Comment:
Micro-classes again …? :D

Reply:
In my opinion, they really make sense here. What would you propose as an alternative?

Reply:
In my opinion, the readability that comes from having a constructor really helps.